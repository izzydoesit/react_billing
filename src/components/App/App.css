/*
COLOR PALETTE: 
- Orange highlight: #f86028
- Teal highlight: #00e3bc
- Dark blue background: #175c6a
- Teal dark highlight: #12c8c1
- Teal light background: #e5fdf9
- Gray border/text: #949598

*/


.App {
  margin: 1%;
  padding: 5%;
  font-family: sans-serif;
}