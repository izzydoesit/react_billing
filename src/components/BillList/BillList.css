div.bill-list-wrapper {
  margin: 1%;
  padding: 1% 1% 1% 0;
}

ul.bill-list {
  padding-left: 0;
  line-height: 1.5em;
  list-style: disc inside url(../../bill-icon-bullet.png);
}