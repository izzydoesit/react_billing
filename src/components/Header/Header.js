// @flow 

import React from 'react';
import './Header.css';

const Header = () =>

  <div className="header-wrapper">
    <header>
      <h1 className="page-title">Welcome to the Clark Take-Home Code Challenge</h1>
    </header>
  </div>

export default Header;
