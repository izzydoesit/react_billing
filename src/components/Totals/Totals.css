div.totals-wrapper {
  margin: 1%;
  padding: 1%;
}

p {
  padding: 2%;
  border: 1px solid #12c8c1;
  background-color: #e5fdf9;
  width: 20%;
  color: #949598
}

.danger {
  border: 2px solid #f86028;
}

.success {
  border: 2px solid #08C6BE;
}

.pending {
  border: 2px solid #FEC87D;
}