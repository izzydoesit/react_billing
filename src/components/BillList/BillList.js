// @flow 

import React, { Component } from 'react';
import Bill from '../Bill/Bill';
import {
  isToday,
  isPastDate,
  isFutureDate,
  formatCentsToDollars } from 'clark-utils';
import type { BillType, BillsType } from '../../types'
import './BillList.css';

type Props = {
  bills?: BillsType,
};

export default class BillList extends Component<Props> {
  static defaultProps = {
    bills: {}
  }

  extractGroup = (groupName: string): Array<BillType> => {
    const bills = this.props.bills;
    let groupList = [];

    if (groupName === 'paid') {
      for (let key in bills) {
        const bill = bills[key];
        if (bill['status'] === 'paid') {
          const billCopy = { ...bill };
          groupList.push(billCopy);
        }
      };
    } else {
      if (groupName === 'outstanding') {
        for (let key in bills) {
          const bill = bills[key];
          if (bill['status'] === 'pending' && (isToday(bill['dueDate']) || isFutureDate(bill['dueDate']))) {
            const billCopy = { ...bill };
            groupList.push(billCopy);
            groupList[groupList.length-1]['status'] = 'outstanding';
          }
        };
      } else if (groupName === 'overdue') {
        for (let key in bills) {
          const bill = bills[key];
          if (bill['status'] === 'pending' && isPastDate(bill['dueDate'])) {
            const billCopy = { ...bill };
            groupList.push(billCopy);
            groupList[groupList.length-1]['status'] = 'overdue';
          }
        };
      }
    }

    return groupList;
  }

  sortGroup = (group: Array<BillType>): Array<BillType> => {
    return group.sort((a, b) => { 
      return (a.dueDate < b.dueDate) ? 1 : ((b.dueDate > a.dueDate) ? -1 : 0);
    });
  }

  render() {
    const paid = this.sortGroup(this.extractGroup('paid'));
    const outstanding = this.sortGroup(this.extractGroup('outstanding'));
    const overdue = this.sortGroup(this.extractGroup('overdue'));
    const sortedBills = [...outstanding, ...overdue, ...paid]

    return (
      <div className="bill-list-wrapper">
        <ul className="bill-list">
          {sortedBills.length === 0
          ? <li className="empty-list-item">No bills have been created...</li>
          : sortedBills.map((bill, index) =>
            <Bill
              key={index}
              id={bill.id}
              status={bill.status}
              amount={formatCentsToDollars(bill.amountInCents)}
              dueDate={bill.dueDate}
            />
          )}
        </ul>
      </div>
    )
  }
}
