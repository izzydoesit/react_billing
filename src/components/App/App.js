// @flow

import React from 'react';
import Header from '../Header/Header'
import BILLS from '../../data';
import Totals from '../Totals/Totals';
import BillList from '../BillList/BillList';
import './App.css';

const App = () =>
  
  <div className="App">
    <Header />
    <Totals bills={BILLS}/>
    <BillList bills={BILLS}/>
  </div>

export default App;
