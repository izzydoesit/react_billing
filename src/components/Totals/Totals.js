// @flow 

import React, { Component } from 'react';
import {
  isToday,
  isPastDate,
  isFutureDate,
  formatCentsToDollars } from 'clark-utils';
import type { BillsType } from '../../types'
import './Totals.css';

type Props = {
  bills: BillsType,
};

export default class Totals extends Component<Props> {
  static defaultProps = {
    bills: {}
  }

  calculateTotals = (bills: BillsType): string => {
    let total = 0, paid = 0, overdue = 0, outstanding = 0;

    for (let key in bills) {
      
      const bill = bills[key];
      const billDueDate = bill['dueDate'];

      if (bill && (bill['status'] === 'pending' && isPastDate(billDueDate))) {
        overdue += bill['amountInCents'];
      } else if (bill['status'] === 'pending' && (isToday(billDueDate) || isFutureDate(billDueDate))) {
        outstanding += bill['amountInCents'];
      } else if (bill['status'] === 'paid') {
        paid += bill['amountInCents'];
      } 
    };

    total = (paid + overdue + outstanding);

    return { total, paid, overdue, outstanding};
  }

  render() {
      const bills = this.props.bills;
      const { total,
              paid, 
              overdue, 
              outstanding} = this.calculateTotals(bills)


    return (
      <div className="totals-wrapper">
        <p className="all-total">Total: <span className="number-total">{formatCentsToDollars(total)}</span></p>
        <p className="paid-total success">Paid: <span className="number-total">{formatCentsToDollars(paid)}</span></p>
        <p className="overdue-total danger">Overdue: <span className="number-total">{formatCentsToDollars(overdue)}</span></p>
        <p className="outstanding-total pending">Outstanding: <span className="number-total">{formatCentsToDollars(outstanding)}</span></p>
      </div>
    )
  }
}
