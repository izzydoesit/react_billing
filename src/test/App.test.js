// @flow

import React from 'react';
import App from '../components/App/App';
import Header from '../components/Header/Header';
import Totals from '../components/Totals/Totals';
import BillList from '../components/BillList/BillList';
import { mount, configure } from 'enzyme';
import { mountToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

describe('<App />', () => {
  let app;

  beforeEach(() => {
    app = mount(<App />);
  });

  it('renders correctly', () => {
    expect(mountToJson(app)).toMatchSnapshot();
  });

  it('renders a div', () => {
    expect(app.find('div.App').length).toEqual(1);
  });

  it('renders a Header component', () => {
    expect(app.find(Header).length).toEqual(1);
  });

  it('renders a Totals component', () => {
    expect(app.find(Totals).length).toEqual(1);
  });

  it('renders a BillList component', () => {
    expect(app.find(BillList).length).toEqual(1);
  });
});
