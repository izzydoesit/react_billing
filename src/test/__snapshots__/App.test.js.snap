// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly 1`] = `
<App>
  <div
    className="App"
  >
    <Header>
      <div
        className="header-wrapper"
      >
        <header>
          <h1
            className="page-title"
          >
            Welcome to the Clark Take-Home Code Challenge
          </h1>
        </header>
      </div>
    </Header>
    <Totals
      bills={
        Object {
          "a": Object {
            "amountInCents": 12300,
            "dueDate": "2018-04-29",
            "id": "a",
            "status": "pending",
          },
          "b": Object {
            "amountInCents": 1153,
            "dueDate": "2017-08-19",
            "id": "b",
            "status": "paid",
          },
          "c": Object {
            "amountInCents": 101363,
            "dueDate": "2018-10-11",
            "id": "c",
            "status": "pending",
          },
          "d": Object {
            "amountInCents": 789127,
            "dueDate": "2018-11-30",
            "id": "d",
            "status": "pending",
          },
          "e": Object {
            "amountInCents": 8129,
            "dueDate": "2018-07-22",
            "id": "e",
            "status": "paid",
          },
          "f": Object {
            "amountInCents": 93020,
            "dueDate": "2018-06-06",
            "id": "f",
            "status": "pending",
          },
          "g": Object {
            "amountInCents": 129308,
            "dueDate": "2018-02-11",
            "id": "g",
            "status": "paid",
          },
          "h": Object {
            "amountInCents": 39333,
            "dueDate": "2018-01-15",
            "id": "h",
            "status": "paid",
          },
          "i": Object {
            "amountInCents": 525222,
            "dueDate": "2018-03-13",
            "id": "i",
            "status": "pending",
          },
        }
      }
    >
      <div
        className="totals-wrapper"
      >
        <p
          className="all-total"
        >
          Total: 
          <span
            className="number-total"
          >
            $16,989.55
          </span>
        </p>
        <p
          className="paid-total success"
        >
          Paid: 
          <span
            className="number-total"
          >
            $1,779.23
          </span>
        </p>
        <p
          className="overdue-total danger"
        >
          Overdue: 
          <span
            className="number-total"
          >
            $6,305.42
          </span>
        </p>
        <p
          className="outstanding-total pending"
        >
          Outstanding: 
          <span
            className="number-total"
          >
            $8,904.90
          </span>
        </p>
      </div>
    </Totals>
    <BillList
      bills={
        Object {
          "a": Object {
            "amountInCents": 12300,
            "dueDate": "2018-04-29",
            "id": "a",
            "status": "pending",
          },
          "b": Object {
            "amountInCents": 1153,
            "dueDate": "2017-08-19",
            "id": "b",
            "status": "paid",
          },
          "c": Object {
            "amountInCents": 101363,
            "dueDate": "2018-10-11",
            "id": "c",
            "status": "pending",
          },
          "d": Object {
            "amountInCents": 789127,
            "dueDate": "2018-11-30",
            "id": "d",
            "status": "pending",
          },
          "e": Object {
            "amountInCents": 8129,
            "dueDate": "2018-07-22",
            "id": "e",
            "status": "paid",
          },
          "f": Object {
            "amountInCents": 93020,
            "dueDate": "2018-06-06",
            "id": "f",
            "status": "pending",
          },
          "g": Object {
            "amountInCents": 129308,
            "dueDate": "2018-02-11",
            "id": "g",
            "status": "paid",
          },
          "h": Object {
            "amountInCents": 39333,
            "dueDate": "2018-01-15",
            "id": "h",
            "status": "paid",
          },
          "i": Object {
            "amountInCents": 525222,
            "dueDate": "2018-03-13",
            "id": "i",
            "status": "pending",
          },
        }
      }
    >
      <div
        className="bill-list-wrapper"
      >
        <ul
          className="bill-list"
        >
          <Bill
            amount="$7,891.27"
            dueDate="2018-11-30"
            id="d"
            key="0"
            status="outstanding"
          >
            <li
              className="bill-item"
            >
              <span
                className="bill-id"
              >
                d
                : 
              </span>
              <span
                className="bill-amount"
              >
                $7,891.27
                , 
              </span>
              <span
                className="bill-status"
              >
                outstanding
                , 
              </span>
              <span
                className="bill-due-date"
              >
                2018-11-30
              </span>
            </li>
          </Bill>
          <Bill
            amount="$1,013.63"
            dueDate="2018-10-11"
            id="c"
            key="1"
            status="outstanding"
          >
            <li
              className="bill-item"
            >
              <span
                className="bill-id"
              >
                c
                : 
              </span>
              <span
                className="bill-amount"
              >
                $1,013.63
                , 
              </span>
              <span
                className="bill-status"
              >
                outstanding
                , 
              </span>
              <span
                className="bill-due-date"
              >
                2018-10-11
              </span>
            </li>
          </Bill>
          <Bill
            amount="$930.20"
            dueDate="2018-06-06"
            id="f"
            key="2"
            status="overdue"
          >
            <li
              className="bill-item"
            >
              <span
                className="bill-id"
              >
                f
                : 
              </span>
              <span
                className="bill-amount"
              >
                $930.20
                , 
              </span>
              <span
                className="bill-status"
              >
                overdue
                , 
              </span>
              <span
                className="bill-due-date"
              >
                2018-06-06
              </span>
            </li>
          </Bill>
          <Bill
            amount="$123.00"
            dueDate="2018-04-29"
            id="a"
            key="3"
            status="overdue"
          >
            <li
              className="bill-item"
            >
              <span
                className="bill-id"
              >
                a
                : 
              </span>
              <span
                className="bill-amount"
              >
                $123.00
                , 
              </span>
              <span
                className="bill-status"
              >
                overdue
                , 
              </span>
              <span
                className="bill-due-date"
              >
                2018-04-29
              </span>
            </li>
          </Bill>
          <Bill
            amount="$5,252.22"
            dueDate="2018-03-13"
            id="i"
            key="4"
            status="overdue"
          >
            <li
              className="bill-item"
            >
              <span
                className="bill-id"
              >
                i
                : 
              </span>
              <span
                className="bill-amount"
              >
                $5,252.22
                , 
              </span>
              <span
                className="bill-status"
              >
                overdue
                , 
              </span>
              <span
                className="bill-due-date"
              >
                2018-03-13
              </span>
            </li>
          </Bill>
          <Bill
            amount="$81.29"
            dueDate="2018-07-22"
            id="e"
            key="5"
            status="paid"
          >
            <li
              className="bill-item"
            >
              <span
                className="bill-id"
              >
                e
                : 
              </span>
              <span
                className="bill-amount"
              >
                $81.29
                , 
              </span>
              <span
                className="bill-status"
              >
                paid
                , 
              </span>
              <span
                className="bill-due-date"
              >
                2018-07-22
              </span>
            </li>
          </Bill>
          <Bill
            amount="$1,293.08"
            dueDate="2018-02-11"
            id="g"
            key="6"
            status="paid"
          >
            <li
              className="bill-item"
            >
              <span
                className="bill-id"
              >
                g
                : 
              </span>
              <span
                className="bill-amount"
              >
                $1,293.08
                , 
              </span>
              <span
                className="bill-status"
              >
                paid
                , 
              </span>
              <span
                className="bill-due-date"
              >
                2018-02-11
              </span>
            </li>
          </Bill>
          <Bill
            amount="$393.33"
            dueDate="2018-01-15"
            id="h"
            key="7"
            status="paid"
          >
            <li
              className="bill-item"
            >
              <span
                className="bill-id"
              >
                h
                : 
              </span>
              <span
                className="bill-amount"
              >
                $393.33
                , 
              </span>
              <span
                className="bill-status"
              >
                paid
                , 
              </span>
              <span
                className="bill-due-date"
              >
                2018-01-15
              </span>
            </li>
          </Bill>
          <Bill
            amount="$11.53"
            dueDate="2017-08-19"
            id="b"
            key="8"
            status="paid"
          >
            <li
              className="bill-item"
            >
              <span
                className="bill-id"
              >
                b
                : 
              </span>
              <span
                className="bill-amount"
              >
                $11.53
                , 
              </span>
              <span
                className="bill-status"
              >
                paid
                , 
              </span>
              <span
                className="bill-due-date"
              >
                2017-08-19
              </span>
            </li>
          </Bill>
        </ul>
      </div>
    </BillList>
  </div>
</App>
`;
